rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,156 @@
+unused code inside the end-user application unless certain scenario happens [2].
+
+As a part of this proposal, there are only two possible options to specify right
+now:
+* Propagation kind: inference or training (the default). Usage example, which
+  enables inference only:
+  - -DNNL_ENABLE_PROP_KIND=INFERENCE
+* Primitives. Usage example, which enables only convolution and matmul:
+  - -DDNNL_ENABLE_PRIMITIVES=CONVOLUTION;MATMUL
+
+This list may be extended in future upon request with data types and, possibly,
+ISA, if it's proven to help with further size reduction with minimal amount of
+enabling efforts and maintenance cost.
+
+Pros:
+1. No API changes and transparent enabling for the end-user.
+2. Simple enabling.
+
+Cons:
+1. Can't be fully automated and relies on the way the code is written and
+   structured.
+2. May require semi-rocket-science validation.
+
+## Additional Comments
+[1] oneDNN v2.2 + gcc 4.8.5 + release mode + standard linker +
+    TBB threading (2017.2) + c++11 + native optimizations on AVX2.
+
+[2] Certain scenario mentioned is the following. Assume linker uses
+implementation list as the primary entry point to fill the binary with
+implementations code. It needs to include all code into the binary which was
+referenced from the list. It finds object files (read, translation units) and
+includes necessary code and it includes some unnecessary, too. Unnecessary code
+can be either of:
+* Instantiations of templated classes.
+* Specializations of templated classes or class members.
+* Non-templated class methods definitions.
+* ... (maybe something else we don't experience in the library) ...
+
+This behavior can't be changed or influenced externally.
+
+## Implementation Details
+
+Main mechanism:
+
+```cmake
+# options.cmake:
+
+foreach(impl in ${DNNL_IMPLS})
+    foreach(prop in ${DNNL_PROPS})
+        string(TOUPPER ${impl} uimpl)
+        string(TOUPPER ${prop} uprop)
+        ...
+        set(BUILD_${uimpl}_${uprop} TRUE)
+    endforeach()
+endforeach()
+```
+
+```cpp
+// src/common/impl_registration.hpp (new header):
+
+#if BUILD_CONVOLUTION_FWD
+#define REGISTER_CONVOLUTION_FWD(...) __VA_ARGS__
+#else
+#define REGISTER_CONVOLUTION_FWD(...)
+#endif
+
+#if BUILD_CONVOLUTION_BWD
+#define REGISTER_CONVOLUTION_BWD(...) __VA_ARGS__
+#else
+#define REGISTER_CONVOLUTION_BWD(...)
+#endif
+```
+
+```cpp
+// src/cpu/cpu_convolution_list.cpp:
+const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    // FWD fp
+    {{forward, f32, f32, f32}, {
+        REGISTER_CONVOLUTION_FWD(CPU_INSTANCE_X64(ip_convolution_fwd_t))
+        ...
+        REGISTER_CONVOLUTION_FWD(CPU_INSTANCE(ref_fused_convolution_fwd_t))
+        nullptr,
+    }},
+```
+
+Additionally, to overcome some of [2], instantiations in separate translations
+units should be wrapped as well:
+
+```cpp
+// src/cpu/ref_convolution.cpp:
+
+REGISTER_CONVOLUTION_FWD(template struct ref_convolution_fwd_t<f32>);
```

Comment:
How is it going to be used for GPU backend? Especially for OCL kernels, as the main impact from OCL impl to binary size comes from kernel strings.

Reply:
Hi Vladimir (by some reason, I can't tag you in here...), thanks for your concern.
I can confirm that the statement about the impact is true. In numbers, it's ~8 MB that are occupied by all kernels if not removed from binary.
So the breakdown for GPU only binary looks like this:
* Core API: ~2 MB.
* String kernels: ~8 MB.
* Implementations: ~1.5-2 MB.
@densamoilov has couple ideas how to reduce the impact from string kernels:
1. To remove the constant overhead coming from preprocessed headers (not sure about the details, but will post them as soon as I have them).
2. To modify generation of kernels in Cmake. Same idea like for CPU, just implemented in a different spot for GPU.
First bullet would help in general, second bullet would utilize the feature proposed. Thanks.

Reply:
@dzarukin 
I'd also propose to remove comments, redundant spaces and empty lines from the kernels strings in the preprocessing script. We did it in clDNN and the improvement was pretty good: https://github.com/openvinotoolkit/openvino/pull/1152

Reply:
Hi @vladimir-paramuzov. So, here's the more detailed breakdown with our findings:
* Full GPU only binary takes ~11 MB.
* Same binary but without any implementations, but with all string kernels takes ~8 MB.
* Same binary but with kernels without headers takes ~3 MB.
* Empty binary with core API takes ~1.8 MB.
Steps 2 and 3 shows that if kernels had a better handling of preprocessed headers, there wouldn't be a 5 MB gap. It also seems that such optimization would be more efficient rather than switching off certain kernels based on selected primitives as suggested initially. So I've created an internal item for GPU team to investigate the issue and provide a technical solution. If, by any reason, additional optimizations would be required, we always may come to the initial proposal and start throwing kernels away based on primitives enabled but I don't expect non-convolutional or non-gemm kernels would take more than third of 2 MB (future rough estimation of kernel space).
Does this sound good for you?

Reply:
@dzarukin Could you include a task for throwing disabled kernels string away into you your backlog at least? So once headers optimization and stuff like that is done, we'll need to estimate binary size impact from useless kernels string. I have a feeling that they can take like 0.5-1MB at least, which is significant for us, since the whole GPU plugin right now takes ~8-9MB, and we'd want onednn to be as small as possible.